File: 백준/Gold/1414. 불우이웃돕기/불우이웃돕기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;

public class Main {
    private static int[] parent;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        PriorityQueue<Edge> pq = new PriorityQueue<>();
        int totalLength = 0;
        for (int i = 0; i < n; i++) {
            char[] line = br.readLine().toCharArray();
            for (int j = 0; j < n; j++) {
                int length = charToLength(line[j]);
                totalLength += length;

                if (i != j && length > 0) {
                    pq.offer(new Edge(i, j, length));
                }
            }
        }

        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        int mstCost = 0;
        int edgeCount = 0;
        while (!pq.isEmpty()) {
            Edge edge = pq.poll();
            if (find(edge.from) != find(edge.to)) {
                union(edge.from, edge.to);
                mstCost += edge.length;
                edgeCount++;
            }
        }

        if (edgeCount == n - 1) {
            System.out.println(totalLength - mstCost);
        } else {
            System.out.println(-1);
        }
    }

    private static int charToLength(char c) {
        if (c >= 'a' && c <= 'z') {
            return c - 'a' + 1;
        } else if (c >= 'A' && c <= 'Z') {
            return c - 'A' + 27;
        }

        return 0;
    }

    private static int find(int x) {
        if (parent[x] == x) return x;

        return parent[x] = find(parent[x]);
    }

    private static void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            parent[rootY] = rootX;
        }
    }

    private static class Edge implements Comparable<Edge> {
        int from, to, length;

        public Edge(int from, int to, int length) {
            this.from = from;
            this.to = to;
            this.length = length;
        }

        @Override
        public int compareTo(Edge o) {
            return Integer.compare(this.length, o.length);
        }
    }
}

File: 백준/Gold/1414. 불우이웃돕기/불우이웃돕기.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.PriorityQueue

private lateinit var parent: IntArray

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()

    val pq = PriorityQueue<Edge>()
    var totalLength = 0
    repeat(n) { i ->
        val line = br.readLine()
        line.forEachIndexed { j, c ->
            val length = charToLength(c)
            totalLength += length

            if (i != j && length > 0) {
                pq.offer(Edge(i, j, length))
            }
        }
    }

    parent = IntArray(n) { it }

    var mstCost = 0
    var edgeCount = 0
    while (pq.isNotEmpty()) {
        val edge = pq.poll()
        if (find(edge.from) != find(edge.to)) {
            union(edge.from, edge.to)
            mstCost += edge.length
            edgeCount++
        }
    }

    if (edgeCount == n - 1) {
        println(totalLength - mstCost)
    } else {
        println(-1)
    }
}

private fun charToLength(c: Char): Int {
    return when (c) {
        in 'a'..'z' -> c - 'a' + 1
        in 'A'..'Z' -> c - 'A' + 27
        else -> 0
    }
}

private fun find(x: Int): Int {
    if (parent[x] == x) return x
    parent[x] = find(parent[x])
    return parent[x]
}

private fun union(x: Int, y: Int) {
    val rootX = find(x)
    val rootY = find(y)

    if (rootX != rootY) {
        parent[rootY] = rootX
    }
}

private data class Edge(val from: Int, val to: Int, val length: Int) : Comparable<Edge> {
    override fun compareTo(other: Edge) = length.compareTo(other.length)
}
